// vue3-baidu-map-gl/es/components/overlay/polygon/index.vue.js
import { defineComponent, inject, watch, provide, nextTick } from 'vue';
import useBaseMapEffect from '../../../hooks/useBaseMapEffect';
import useLifeCycle from '../../../hooks/useLifeCycle';
import { callWhenDifferentValue, bindEvents } from '../../../utils';

const __default__ = {
    name: 'BPolygon'
};
var script = /*#__PURE__*/ defineComponent(Object.assign(Object.assign({}, __default__), { props: {
        path: { type: Array, required: true },
        isBoundary: { type: Boolean, required: false },
        strokeColor: { type: String, required: false, default: '#000' },
        autoCenter: { type: Boolean, required: false, default: true },
        strokeWeight: { type: Number, required: false, default: 2 },
        strokeOpacity: { type: Number, required: false, default: 1 },
        strokeStyle: { type: null, required: false, default: 'solid' },
        fillColor: { type: String, required: false, default: '#fff' },
        fillOpacity: { type: Number, required: false, default: 0.3 },
        enableMassClear: { type: Boolean, required: false },
        enableEditing: { type: Boolean, required: false },
        enableClicking: { type: Boolean, required: false },
        geodesic: { type: Boolean, required: false, default: false },
        clip: { type: Boolean, required: false, default: true },
        onClick: { type: null, required: false },
        onDblclick: { type: null, required: false },
        onMousedown: { type: null, required: false },
        onMouseup: { type: null, required: false },
        onMouseout: { type: null, required: false },
        onMouseover: { type: null, required: false },
        onRemove: { type: null, required: false },
        onLineupdate: { type: null, required: false }
    }, emits: [
        'initd',
        'unload',
        'click',
        'dblclick',
        'mousedown',
        'mouseup',
        'mouseout',
        'mouseover',
        'remove',
        'lineupdate'
    ], setup(__props, { emit: vueEmits }) {
        const props = __props;
        const { ready } = useLifeCycle();
        const injectBaseMapSetCenterAndZoom = inject('baseMapSetCenterAndZoom');
        let polygon;
        useBaseMapEffect((map) => {
            const cal = () => {
                polygon && map.removeOverlay(polygon);
            };
            const init = () => {
                if (!props.path || !props.path.length)
                    return;
                const { path, strokeColor, strokeWeight, strokeOpacity, strokeStyle, fillOpacity, fillColor, enableMassClear, enableEditing, enableClicking, geodesic, clip, isBoundary } = props;
                const pathPoints = isBoundary ? path : pathPointsToMapPoints(path);
                if (!pathPoints)
                    return;
                polygon = new BMapGL.Polygon(pathPoints, {
                    strokeColor,
                    strokeWeight,
                    strokeOpacity,
                    strokeStyle,
                    fillOpacity,
                    fillColor,
                    enableMassClear,
                    enableEditing,
                    enableClicking,
                    geodesic,
                    clip
                });
                map.addOverlay(polygon);
                bindEvents(props, vueEmits, polygon);
                ready(map, polygon);
                syncMapCenter();
                watch(() => props.strokeColor, setStrokeColor);
                watch(() => props.strokeOpacity, setStrokeOpacity);
                watch(() => props.fillColor, setFillColor);
                watch(() => props.fillOpacity, setFillOpacity);
                watch(() => props.strokeWeight, setStrokeWeight);
                watch(() => props.strokeStyle, setStrokeStyle);
                watch(() => props.enableMassClear, setMassClear);
                watch(() => props.enableEditing, setEditing);
            };
            init();
            // 监听值变化, 初始为空时不会初始化, 不为空值时初始化
            watch(() => props.path, callWhenDifferentValue((path) => {
                polygon ? setPath(path) : init();
            }), {
                deep: true
            });
            return cal;
        });
        provide('getOverlayInstance', () => polygon);
        function syncMapCenter() {
            nextTick(() => {
                var _a;
                // 自动设置中心点
                if (props.autoCenter) {
                    // 获取中心点
                    try {
                        const center = (_a = polygon.getBounds()) === null || _a === void 0 ? void 0 : _a.getCenter();
                        injectBaseMapSetCenterAndZoom(center);
                    }
                    catch (e) {
                        console.warn('set center error', e);
                    }
                }
            });
        }
        function pathPointsToMapPoints(pathPoints) {
            return pathPoints.map(({ lng, lat }) => new BMapGL.Point(lng, lat));
        }
        function setPath(path) {
            if (props.isBoundary) {
                polygon.setPath(path);
            }
            else {
                polygon.setPath(pathPointsToMapPoints(path));
            }
            syncMapCenter();
        }
        function setStrokeColor(color) {
            polygon.setStrokeColor(color);
        }
        function setFillColor(color) {
            polygon.setFillColor(color);
        }
        function setStrokeOpacity(opacity) {
            polygon.setStrokeOpacity(opacity);
        }
        function setFillOpacity(opacity) {
            polygon.setFillOpacity(opacity);
        }
        function setStrokeWeight(weight) {
            polygon.setStrokeWeight(weight);
        }
        function setStrokeStyle(style) {
            polygon.setStrokeStyle(style);
        }
        function setMassClear(enableMassClear) {
            enableMassClear ? polygon.enableMassClear() : polygon.disableMassClear();
        }
        function setEditing(enableEditing) {
            enableEditing ? polygon.enableEditing() : polygon.disableEditing();
        }
        return (_ctx, _cache) => {
            return null;
        };
    } }));

script.__file = "packages/components/overlay/polygon/index.vue";

export { script as default };
