// vue3-baidu-map-gl/es/components/overlay/label/index.vue.js
import { defineComponent, watch, provide } from 'vue';
import useBaseMapEffect from '../../../hooks/useBaseMapEffect';
import useLifeCycle from '../../../hooks/useLifeCycle';
import { callWhenDifferentValue, isDef, bindEvents } from '../../../utils/index';

const __default__ = {
    name: 'BLabel'
};
var script = /*#__PURE__*/ defineComponent(Object.assign(Object.assign({}, __default__), { props: {
        content: { type: String, required: true },
        position: { type: null, required: true },
        offset: { type: Object, required: false, default: () => ({
                x: 0,
                y: 0
            }) },
        style: { type: null, required: false },
        enableMassClear: { type: Boolean, required: false, default: true },
        onClick: { type: null, required: false },
        onDblclick: { type: null, required: false },
        onMousedown: { type: null, required: false },
        onMouseup: { type: null, required: false },
        onMouseout: { type: null, required: false },
        onMouseover: { type: null, required: false },
        onRemove: { type: null, required: false },
        onRightclick: { type: null, required: false }
    }, emits: [
        'initd',
        'unload',
        'click',
        'dblclick',
        'mousedown',
        'mouseup',
        'mouseout',
        'mouseover',
        'remove',
        'rightclick'
    ], setup(__props, { emit: vueEmits }) {
        const props = __props;
        const { ready } = useLifeCycle();
        let label;
        useBaseMapEffect((map) => {
            const cal = () => {
                label && map.removeOverlay(label);
            };
            const init = () => {
                const { content, position, offset, enableMassClear, style } = props;
                const options = {
                    position: new BMapGL.Point(position.lng, position.lat),
                    offset: new BMapGL.Size(offset.x, offset.y),
                    enableMassClear
                };
                label = new BMapGL.Label(content, options);
                // 自定义文本标注样式
                if (style) {
                    label.setStyle(style);
                }
                map.addOverlay(label);
                // isDef(zIndex) && setZIndex(zIndex)
                bindEvents(props, vueEmits, label);
            };
            init();
            ready(map, label);
            // 监听值变化
            watch(() => props.position, callWhenDifferentValue(setPosition), { deep: true });
            watch(() => props.offset, callWhenDifferentValue(setOffset), { deep: true });
            watch(() => props.style, callWhenDifferentValue(setStyle), { deep: true });
            watch(() => props.content, setContent);
            watch(() => props.enableMassClear, setMassClear);
            return cal;
        });
        provide('getOverlayInstance', () => label);
        // function setZIndex(zIndex: number) {
        //   console.log('shezhi ')
        //   label.setZIndex(zIndex)
        // }
        function setPosition(position) {
            label.setPosition(new BMapGL.Point(position.lng, position.lat));
        }
        function setStyle(styles) {
            if (isDef(styles))
                label.setStyle(styles);
        }
        function setContent(content) {
            label.setContent(content);
        }
        function setOffset(offset) {
            label.setOffset(new BMapGL.Size(offset.x, offset.y));
        }
        function setMassClear(enableMassClear) {
            enableMassClear ? label.enableMassClear() : label.disableMassClear();
        }
        return (_ctx, _cache) => {
            return null;
        };
    } }));

script.__file = "packages/components/overlay/label/index.vue";

export { script as default };
