// vue3-baidu-map-gl/es/components/layer/district-layer/index.vue.js
import { defineComponent } from 'vue';
import useLifeCycle from '../../../hooks/useLifeCycle';
import useBaseMapEffect from '../../../hooks/useBaseMapEffect';
import { warn, bindEvents } from '../../../utils';

const __default__ = {
    name: 'BDistrictLayer'
};
var script = /*#__PURE__*/ defineComponent(Object.assign(Object.assign({}, __default__), { props: {
        name: { type: String, required: true },
        kind: { type: null, required: false, default: 0 },
        fillColor: { type: String, required: false, default: '#fdfd27' },
        fillOpacity: { type: Number, required: false, default: 1 },
        strokeColor: { type: String, required: false, default: '#231cf8' },
        viewport: { type: Boolean, required: false, default: false },
        onClick: { type: null, required: false },
        onDblclick: { type: null, required: false },
        onRightclick: { type: null, required: false },
        onRightdblclick: { type: null, required: false },
        onMousemove: { type: null, required: false },
        onMouseover: { type: null, required: false },
        onMouseout: { type: null, required: false }
    }, emits: ['initd', 'unload', 'mouseover', 'mouseout', 'click'], setup(__props, { emit: vueEmits }) {
        const props = __props;
        const { ready } = useLifeCycle();
        let districtLayer;
        useBaseMapEffect((map) => {
            if (!props.name)
                return warn('DistrictLayer props.name is required');
            const { name, kind, fillColor, fillOpacity, strokeColor, viewport } = props;
            districtLayer = new BMapGL.DistrictLayer({
                name: `(${name})`,
                kind,
                fillColor,
                fillOpacity,
                strokeColor,
                viewport
            });
            map.addDistrictLayer(districtLayer);
            bindEvents(props, vueEmits, districtLayer);
            ready(map, districtLayer);
            return () => map.removeDistrictLayer(districtLayer);
        });
        return (_ctx, _cache) => {
            return null;
        };
    } }));

script.__file = "packages/components/layer/district-layer/index.vue";

export { script as default };
